### ISSUE TITLE
Bug: more-route-selector don't work this dynamic core-item made with template repeat

### ISSUE THREAD

**Issue by jmorille** [Polymore/more-routing#17]
Replace this working static menu
```javascript
 <more-route-selector>
      <core-menu>
        <core-item  icon="home"  label="Home" route="home">
          <a href="{{urlFor('home')}}"></a>
        </core-item>
        <core-item icon="supervisor-account"  label="User" route="user">
          <a href="{{urlFor('user')}}"></a>
        </core-item>
        <core-item icon="perm-media" label="Portfolio" route="portfolio">
          <a href="{{urlFor('portfolio')}}"></a>
        </core-item>
        <core-item icon="perm-contact-cal" label="Contact" route="contact">
          <a href="{{urlFor('contact')}}"></a>
        </core-item>
      </core-menu>
    </more-route-selector>
```

with the equivalent dynamic one
```javascript
    <more-route-selector>
      <core-menu>
        <template repeat="{{menu in menus}}">
          <core-item icon="{{menu.icon}}" label="{{menu.label}}" route="{{menu.route}}">
            <a href="{{urlFor(menu.route)}}"></a>
          </core-item>
        </template>
      </core-menu>
    </more-route-selector>
</template>
<script>
..
 menus: undefined,
      created: function () {
        this.menus = [
          {icon: 'home', label: 'Home', route: 'home'},
          {icon: 'supervisor-account', label: 'User', route: 'user'},
          {icon: 'perm-media', label: 'Portfolio', route: 'portfolio'},
          {icon: 'perm-contact-cal', label: 'Contact', route: 'contact'}
        ];
      }
```
and no selection was possible on the menu item with the template one.

**Comment by jmorille** [comment]
In the  more-route-selector.html:90
this.$.selection.routes have the routes
 this.$.selection.routes: Array[5]
0: null
1: Route
_paramObserver: ObjectObserver
 **active: true **
children: Array[0]
compiled: Array[0]
depth: 0
driver: HashDriver
fullPath: "/"
numParams: 0
params: Object
parent: null
parts: Array[1]
path: "/"
__proto__: Route
2: Route
_paramObserver: ObjectObserver
 **active: true **
children: Array[0]
compiled: Array[1]
depth: 1
driver: HashDriver
fullPath: "/user"
numParams: 0
params: Object
parent: null
parts: Array[1]
path: "/user"
__proto__: Route
3: Route
_paramObserver: ObjectObserver
active: false
children: Array[0]
compiled: Array[1]
depth: 1
driver: HashDriver
fullPath: "/portfolio"
numParams: 0
params: Object
parent: null
parts: Array[1]
path: "/portfolio"
__proto__: Route
4: Route
_paramObserver: ObjectObserver
active: false
children: Array[0]
compiled: Array[1]
depth: 1
driver: HashDriver
fullPath: "/contact"
numParams: 0
params: Object
parent: null
parts: Array[1]
path: "/contact"
__proto__: Route
length: 5
__proto__: Array[0]

**Comment by jmorille** [comment]
The selected menu is the next one that should be selected.
It the selection is based on the index of the  this.$.selection.routes array,
the idx 0, generated by the template could explit the wrong menu selected ??

**Comment by jmorille** [comment]
If you call the route http://localhost:9000/#!/user
the code below is call 2 times, one with  selectedIndex = 2 and next with selectedIndex = 1

```javascript
 selectedIndexChanged: function() {
        if (!this.routingIsReady || !this._managedSelector) return;
        this._managedSelector.selected = this.selectedIndex; 
      },
```

**Comment by nevir** [comment]
Ah, thanks for digging into this! 

The current assumption in `<more-route-selector>` is definitely that the children are static. Currently, it's only updating the [route selection in `domReady`](https://github.com/Polymore/more-routing/blob/master/more-route-selector.html#L90).

As a complete and utter hack for now, try updating the `<more-route-selector>`'s internal `<more-route-selection>` after the template has been stamped out (e.g. `routeSelector.$.selection.routes = routeSelector._routesFromChildren()` in your element's `ready`, or async after that)

I'm unfortunately not going to have too much bandwidth to fix this for a week or two though :( (patches welcome!)

**Comment by nevir** [comment]
Aha! So the null route is coming from the template element (template repeats stay in the final DOM, and all their generated elements follow them) - that's also where you're getting the warning. Might be worth having `<more-route-selector>` ignore `<template>` elements.

On the two routes being active, that's correct and has been a difficult behavior of more routing for me to document :( Here's a quick try:

If your URL is `/foo/bar`, the route `/foo/bar` is active, but the route `/foo` would also be active, as is the route `/`. (They're a prefix match). This allows for better composition where elements/contexts can create sub routes.

`<more-route-selector>` will choose the _most specific_ route (e.g. route w/ the most path parts). So if `/` and `/foo/` are active, it'll choose `/foo` over `/`. The bulk of that logic is in https://github.com/Polymore/more-routing/blob/master/more-route-selection.html#L141-171 when it sorts the routes by specificity

**Comment by nevir** [comment]
There's definitely some issues here, though I probably won't be able to dig into them for a week or two :( (patches welcome, though, if you do manage to root it out!)

I suspect the `null` is confusing things, and that double call to `selectedIndexChanged` also seems suspect...

**Comment by jmorille** [comment]
My feeling for this bug, is that caused by a difference of selectedIndex value of the 2 components *<more-route-selector>* and the *<core-pages>*.
The  <template> element included a null node in this.$.selection.routes that create the +1 shift in the selected nodes. It could explain why the second selectedIndexChanged event that fire
I need more time to revert your code, and for confirm my feeling.

**Comment by nevir** [comment]
Yeah, apologies for it being rough, and thanks for digging into this

**Comment by jmorille** [comment]
No problem, it was not upset.. no problem, I am ok
I past an interesting time to revert you code and learn some good hidden Polymer's undocumented technic. Polymer is a really good jobs by the way (even if I was not sure about the apis change of the 0.8) and I am sure that will change the web as you know it.

But like you, I have a little busy too, and I not sure to have time to digging into this,  this week.

**Comment by nevir** [comment]
No worries!

And yeah, this will likely change a fair bit for 0.8. I will likely be not using `Object.observe` once I port this over to 0.8, which should simplify `more-routing`.

**Comment by sgoryachkin** [comment]
I have the same problem with more-router 1.0
in function _updateRoutes: this._managedSelector.items is empty during the more-route-selector is attached, If core-item made with template repeat.

**Comment by sgoryachkin** [comment]
It is iron-selector problem in my case:
https://github.com/PolymerElements/iron-selector/issues/89
